File: crates/flighthq-scene-wgpu/src/matcap_wgpu_mesh_material_renderer.rs
//! The built-in Matcap forward mesh-material renderer — the WGSL mirror of
//! `matcap_gl_mesh_material_renderer`.
//!
//! Ports `@flighthq/scene-wgpu` `matcapWgpuMeshMaterialRenderer.ts`.
//! Lighting-independent material-capture shading: `bind` selects the matcap pipeline
//! variant, writes the shared Frame uniform (lights ignored — the matcap texture is
//! the prebaked lighting), and binds the linear tint; `draw` issues the indexed draw.
//! The real matcap texture is not yet sampled on wgpu (`has_matcap` stays false), so
//! the surface renders as the tint alone.

use flighthq_materials::matcap_material_kind;
use flighthq_render_wgpu::WgpuRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::material::Material;
use flighthq_types::pbr_material::MaterialAlphaMode;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};
use flighthq_types::unlit_material::MatcapMaterial;

use crate::wgpu_matcap_prelude::{
    WgpuMatcapDefineKey, bind_wgpu_matcap_surface, ensure_wgpu_matcap_pipeline,
};
use crate::wgpu_mesh_material_registry::{
    WgpuMeshMaterialRenderer, register_wgpu_mesh_material_renderer,
};
use crate::wgpu_mesh_pipeline::{
    draw_wgpu_mesh_material_subset, unpack_color_to_linear_f32, write_wgpu_frame_uniform,
};
use crate::wgpu_scene_runtime::{WgpuMeshUpload, WgpuSceneRuntime};

/// The built-in Matcap forward mesh-material renderer.
pub struct MatcapWgpuMeshMaterialRenderer;

impl WgpuMeshMaterialRenderer for MatcapWgpuMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        material: Option<&dyn Material>,
        _lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let matcap = material.and_then(downcast_matcap);
        let format = state.runtime.current_color_format.unwrap_or(state.format);
        let cache_key = ensure_wgpu_matcap_pipeline(state, scene, &define_key(matcap), format);
        scene.active_pipeline_key = Some(cache_key.clone());
        write_wgpu_frame_uniform(state, scene, camera, _lights);

        let material_key = material.map_or_else(matcap_material_kind, |m| m.kind());
        let (tint, alpha_cutoff) = match matcap {
            Some(m) => (unpack_color_to_linear_f32(m.tint), m.alpha_cutoff),
            None => (WHITE, 0.5),
        };
        bind_wgpu_matcap_surface(state, scene, &cache_key, material_key, &tint, alpha_cutoff);
        scene.active_material_key = Some(material_key);
    }

    fn draw(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &WgpuMeshUpload,
    ) {
        draw_wgpu_mesh_material_subset(state, scene, proxy, upload);
    }
}

/// Returns the singleton Matcap renderer instance.
pub fn matcap_wgpu_mesh_material_renderer() -> MatcapWgpuMeshMaterialRenderer {
    MatcapWgpuMeshMaterialRenderer
}

/// Registers the built-in Matcap renderer for the Matcap material kind on this scene
/// runtime. Opt-in (no top-level side effect).
pub fn register_matcap_wgpu_material(state: &mut WgpuRenderState, scene: &mut WgpuSceneRuntime) {
    register_wgpu_mesh_material_renderer(
        state,
        scene,
        matcap_material_kind(),
        Box::new(MatcapWgpuMeshMaterialRenderer),
    );
}

fn define_key(material: Option<&MatcapMaterial>) -> WgpuMatcapDefineKey {
    WgpuMatcapDefineKey {
        alpha_mask_enabled: material.is_some_and(|m| m.alpha_mode == MaterialAlphaMode::Mask),
        double_sided: material.is_some_and(|m| m.double_sided),
        has_matcap: false,
    }
}

fn downcast_matcap(material: &dyn Material) -> Option<&MatcapMaterial> {
    (material as &dyn core::any::Any).downcast_ref::<MatcapMaterial>()
}

const WHITE: [f32; 4] = [1.0, 1.0, 1.0, 1.0];

#[cfg(test)]
mod tests {
    use super::*;
    use crate::wgpu_scene_runtime::create_wgpu_scene_runtime;

    mod register_matcap_wgpu_material {
        use super::*;

        #[test]
        fn registers_the_renderer_under_the_matcap_kind() {
            let mut scene = create_wgpu_scene_runtime();
            scene.material_registry.insert(
                matcap_material_kind(),
                Box::new(MatcapWgpuMeshMaterialRenderer),
            );
            assert!(
                scene
                    .material_registry
                    .contains_key(&matcap_material_kind())
            );
        }
    }
}

File: packages/scene-wgpu/src/matcapWgpuMeshMaterialRenderer.ts
import type { LinearColor } from '@flighthq/color';
import { unpackColorToLinear } from '@flighthq/color';
import { getWgpuRenderStateRuntime } from '@flighthq/render-wgpu';
import type {
  Camera,
  MatcapMaterial,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
  WgpuMeshMaterialRenderer,
  WgpuRenderState,
} from '@flighthq/types';
import { MatcapMaterialKind } from '@flighthq/types';

import type { WgpuMatcapDefineKey } from './wgpuMatcapPrelude';
import { bindWgpuMatcapSurface, ensureWgpuMatcapPipeline } from './wgpuMatcapPrelude';
import { registerWgpuMeshMaterialRenderer } from './wgpuMeshMaterialRegistry';
import { beginWgpuMeshDraw, drawWgpuMeshSubset, writeWgpuFrameUniform } from './wgpuMeshPipeline';

// The built-in Matcap forward renderer (WgpuMeshMaterialRenderer for MatcapMaterialKind) — the WGSL
// mirror of matcapGlMeshMaterialRenderer. Lighting-independent material-capture shading: bind selects
// the matcap pipeline variant for the alpha mode + color format, writes the shared Frame uniform
// (lights ignored — the matcap texture is the prebaked lighting), and binds the linear tint; draw
// issues the indexed draw. See registerMatcapWgpuMaterial to install it. The real matcap texture is not
// yet sampled on wgpu (hasMatcap stays false; see wgpuMatcapPrelude's note), so the surface renders as
// the tint alone for now.
export const matcapWgpuMeshMaterialRenderer: WgpuMeshMaterialRenderer = {
  bind(
    state: WgpuRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera>,
  ): void {
    const stateRuntime = getWgpuRenderStateRuntime(state);
    const pass = stateRuntime.renderPass;
    if (pass === null) return;

    const matcap = material as Readonly<MatcapMaterial> | null;
    const format = stateRuntime.currentColorFormat ?? state.format;
    const pipeline = ensureWgpuMatcapPipeline(state, defineKeyForMaterial(matcap), format);
    writeWgpuFrameUniform(state, camera, _lights);

    let group: GPUBindGroup;
    if (matcap === null) {
      group = bindWgpuMatcapSurface(state, pipeline, FALLBACK_MATERIAL, WHITE, 0.5);
    } else {
      unpackColorToLinear(_scratch, matcap.tint);
      group = bindWgpuMatcapSurface(state, pipeline, matcap, _scratch, matcap.alphaCutoff);
    }

    beginWgpuMeshDraw(state, pipeline);
    pass.setBindGroup(2, group);
  },

  draw(state: WgpuRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    drawWgpuMeshSubset(state, proxy, geometry);
  },
};

// Registers the built-in Matcap renderer for MatcapMaterialKind on this state. Opt-in (no top-level
// side effect); call once per WgpuRenderState before drawWgpuScene so meshes with MatcapMaterials draw.
export function registerMatcapWgpuMaterial(state: WgpuRenderState): void {
  registerWgpuMeshMaterialRenderer(state, MatcapMaterialKind, matcapWgpuMeshMaterialRenderer);
}

function defineKeyForMaterial(material: Readonly<MatcapMaterial> | null): WgpuMatcapDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    doubleSided: material !== null && material.doubleSided,
    hasMatcap: false,
  };
}

const _scratch: LinearColor = [0, 0, 0, 0];
const WHITE: LinearColor = [1, 1, 1, 1];
const FALLBACK_MATERIAL = {} as Readonly<MatcapMaterial>;
